src/processor.cpp

```diff
@@ -0,0 +1,401 @@
+{
+    mandatory( jo, was_loaded, "name", name );
+    //TODO: modify to allow lists of components
+    mandatory( jo, was_loaded, "components", components,  string_id_reader<itype_id> {});
+    mandatory( jo, was_loaded, "output", output );
+    mandatory( jo, was_loaded, "fuel_intake", fuel_intake );
+    mandatory( jo, was_loaded, "duration", duration );
+}
+
+void process_data::check() const
+{
+    const process_id proc_id( id.str() );
+
+    if( !proc_id.is_valid() ) {
+        debugmsg( "There is no process with the name %s", id.c_str() );
+    if( !components.is_valid() ) {
+        debugmsg( "Invalid components \"%s\" in \"%s\".", components.c_str(), id.c_str() );
+    }
+    if( !output.is_valid() ) {
+        debugmsg( "Invalid output \"%s\" in \"%s\".", output.c_str(), id.c_str() );
+    }
+    if( !fuel_intake.is_valid() ) {
+        debugmsg( "Invalid fuel intake \"%s\" in \"%s\".", fuel_intake.c_str(), id.c_str() );
+    }
+    if( !duration.is_valid() ) {
+        debugmsg( "Invalid duration \"%s\" in \"%s\".", duration.c_str(), id.c_str() );
+    }
+}
+
+void process_data::load( JsonObject &jo, const std::string &src )
+{
+    processes_data.load( jo, src );
+}
+
+void process_data::check()
+{
+    processes_data.check();
+}
+
+void process_data::reset()
+{
+    processes_data.reset();
+}
+
+void interact_with_processor( const tripoint &examp, player &p )
+{
+    //Get processor id at examp
+    processor_id pid = get_processor_id( examp );
+    //Do I need the debugmsg-es here or do they get checked in process_data::load?
+    if( !processors_data.is_valid( pid ) ) {
+        debugmsg( "Examined %s has action interact_with_processor, but has no processor data associated", g->m.furn( examp ).id().c_str() );
+        return;
+    }
+    //Get data associated with said processor id
+    const processor_data &processor = processors_data.obj( pid );
+    std::vector<auto> possible_processes;
+    for( auto &p : processor.processes() ) {
+        if( p.is_valid() ){
+            possible_processes.push_back( process_data.obj( process_id( p ) ) );
+        } else {
+            //Debugmsg probably not exactly right,
+            //I'm trying to return the processor name and the problematic process name too
+            debugmsg( "Examined %s has non-valid process data %s2 associated", processor.id().c_str(), get_process_id( p ).c_str() );
+            return;
+        }
+    }
+
+    //  Figure out which process to run, easier to work with than
+    //  trying to figure out desired process from items on tile,
+    //  Can become a mess with processor with multiple possible processes
+    std::vector<std::string> p_names;
+    for( auto p : possible_processes ){
+        p_names.push_back( p.name );
+    }
+    int p_index = 0;
+    if( p_names.size() > 1 ){
+        p_names.push_back(_("Cancel"));
+        p_index = menu_vec(true, _("Possible processes"), p_names);
+        if (p_index == (int)p_names.size() - 1) {
```

Comment:
You also need to handle results `<0`, because cancelling out of a menu does that.